rewrite coap gateway (#1824)
emqtt/emq-docs-cn: en_US/gateway/coap.md

To ensure secure communication, the CoAP gateway includes the **Connection Mode** feature which provides client access authentication. This feature restricts unauthorized CoAP clients from accessing the system, thus enhancing security and preventing potential attacks. 

Below is the feature list supported in connection mode and connectionless mode.

| Feature           | Connectionless Mode | Connection Mode |
| ----------------- | ------------------- | --------------- |
| Message Publish   | √                   | √               |
| Topic Subscribe   | √                   | √               |
| Topic Unsubscribe | √                   | √               |
| Create connection | ×                   | √               |
| Close connection  | ×                   | √               |
| Heartbeat         | ×                   | √               |
| Authentication    | ×                   | √               |

<!--a brief introduction of the architecture-->

## Enable the CoAP Gateway

> In EMQX 5.0, CoAP gateways can be configured and enabled through the Dashboard, HTTP API, and configuration file `emqx.conf`. This section takes the configuration via Dashboard as an example to illustrate the operating steps. 

Reviewer: typo?
CoAP gateways -> CoAP gateway

> In EMQX 5.0, CoAP gateways can be configured and enabled through the Dashboard, HTTP API, and configuration file `emqx.conf`. This section takes the configuration via Dashboard as an example to illustrate the operating steps. 

Author: ```suggestion
In EMQX 5.0, CoAP gateway can be configured and enabled through the Dashboard, HTTP API, and configuration file `emqx.conf`. This section takes the configuration via Dashboard as an example to illustrate the operating steps. 
```

## Customize Your CoAP Gateway

In addition to the default settings, EMQX provides a variety of configuration options to better accommodate your specific business requirements. This section offers an in-depth overview of the various fields available on the **Gateways** page. See the texts below the screenshot for a comprehensive explanation of each field. 

<img src="./assets/coap-basic-conf.png" alt="image-20230420152920254" style="zoom:50%;" />

- **Connection Required**: Set whether to enable connectionless or connection mode, default: **false** (connectionless), optional values: **false** (connectionless), **true** (connection).

- **Notification Message Type**: Set the type of CoAP messages to be delivered, default: **qos**; optional values: 

  - **qos**: Whether the CoAP notification needs to be acknowledged depends on the QoS level of the received message, 
    - QoS 0, no acknowledgment is required from the client, 
    - QoS 1/2, acknowledgment is required from the client. 
  - **con**: The CoAP notification should be acknowledged by the client. 

>   - **non**: The CoAP Notification need not be acknowledged by the client. 

Reviewer: ```suggestion
  - **non**: The CoAP notification need not be acknowledged by the client. 
```

- **Heatbeat**: Only needed if **Connection Required** is set to **true**, set the minimum heartbeat interval to keep the connection alive; default: 30s. 

- **Enable Statistics**: Set whether to allow the Gateway to collect and report statistics; default: **true**, optional values: **true**, **false**.

- **Subscriber QoS**: Set the default QoS level for subscribe requests, default: **coap**, optional values:

  - **coap**: Follow the setting in **Notification Message Type** to determine the QoS level
    - Use QoS 0 if no acknowledgment is required.
    - Use QoS 1 if acknowledgment is required.
  - **qos0**, **qos1**, **qos2**

- **Publish QoS**: Set the default QoS level for publish requests, default: **coap**, optional values: **coap**, **qos0**, **qos1**, **qos2**; the values here 

> - **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different listeners, for example, *mqttsn*.

Reviewer: ```suggestion
- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different protocols, for example, *MQTT-SN*.
```

>   **Note**: The prefixed string will be removed from the topic name when the message is delivered to the subscriber. 

Reviewer: It is more accurate, I think:

This topic prefix is managed by the gateway. CoAP clients do not need to add this prefix explicitly when publishing and subscribing.

>   **Note**: The prefixed string will be removed from the topic name when the message is delivered to the subscriber. 

Author: ```suggestion
  **Note**: This topic prefix is managed by the gateway. CoAP clients do not need to add this prefix explicitly when publishing and subscribing.
```

### Add Listeners 

> By default, one UPD listener with the name of **default** is already configured on port `5683`, which supports up to 1,024,000 concurrent connections. You can click **Settings** for more customized settings for **Delete** to delete the listener. Or click **Add Listener** to add a new listener.

Reviewer: Typos:
UPD listener -> UDP listener

> By default, one UPD listener with the name of **default** is already configured on port `5683`, which supports up to 1,024,000 concurrent connections. You can click **Settings** for more customized settings for **Delete** to delete the listener. Or click **Add Listener** to add a new listener.

Author: ```suggestion
By default, one UDP listener with the name of **default** is already configured on port `5683`, which supports up to 1,024,000 concurrent connections. You can click **Settings** for more customized settings for **Delete** to delete the listener. Or click **Add Listener** to add a new listener.
```

> <img src="/Users/lena/Documents/GitHub/rebalancing/emqx-docs/en_US/gateway/assets/mqttsn-listerner.png" alt="MQTTSN listener" style="zoom:50%;" />

Reviewer: Invalid image path

> <img src="/Users/lena/Documents/GitHub/rebalancing/emqx-docs/en_US/gateway/assets/mqttsn-listerner.png" alt="MQTTSN listener" style="zoom:50%;" />

Author: ```suggestion
<img src="./assets/mqttsn-listerner.png" alt="MQTTSN listener" style="zoom:50%;" />
```

  "user_id_type": "username"
}'
```

::: tip

Unlike the MQTT protocol, **the gateway only supports the creation of an authenticator, not a list of authenticators (or an authentication chain)**. If no authenticator is enabled, all CoAP clients are allowed to log in.

:::

> ## Reference: CoAP Client Guide in Connection Mode

Author: ```suggestion
## Reference: CoAP Client Guide
```
